crates/relayer/src/worker/packet.rs

```diff
@@ -151,6 +205,133 @@ fn handle_packet_cmd<ChainA: ChainHandle, ChainB: ChainHandle>(
+            }
+            // It is not authorized to filter WriteAcknowledgement at the moment.
+            // This is because in order to filter them the worker would need to
+            // query the IncentivizedPacket events seen by the worker handling packets
+            // from ChainB to ChainA or it would need to query ChainA for incentivized
+            // packets.
+            // In addition if the WriteAcknowledgment are not relayed, no fees will be paid.
+            //IbcEvent::WriteAcknowledgement(ack) => get_incentivized_for_write_acknowledgement(link, ack, event.height.revision_height(), incentivized_ack_cache.clone()),
+        }
+        let filtered_batch = filter_batch(batch, incentivized_recv_cache, fee_filter);
+        handle_update_schedule(link, 0, path, filtered_batch)
+    } else {
+        Ok(())
+    }
+}
+
+/// It is only possible to filter incentivized packets using the `recv_fee` as criteria.
+/// This method is a helper which can be used if filtering using `ack_fee` is added.
+/// It queries the chain for incentivized packet information regarding the `WriteAcknowledgement`,
+/// and stores this information in a Read/Write Cache.
+fn _get_incentivized_for_write_acknowledgement<ChainA: ChainHandle, ChainB: ChainHandle>(
+    link: &mut Link<ChainA, ChainB>,
+    ack: WriteAcknowledgement,
+    height: u64,
+    incentivized_ack_cache: RwArc<Cache<Sequence, IdentifiedPacketFees>>,
+) {
+    let dst_chain = link.a_to_b.dst_chain();
+
+    // Build PacketId required for the QueryIncentivizedPacketRequest
+    let packet_id = PacketId {
+        port_id: ack.packet.source_port.to_string(),
+        channel_id: ack.packet.source_channel.to_string(),
+        sequence: ack.packet.sequence.into(),
+    };
+
+    let request = QueryIncentivizedPacketRequest {
+        packet_id: Some(packet_id),
+        query_height: height,
+    };
+
+    match dst_chain.query_incentivized_packet(request) {
+        Ok(ev) => {
+            if let Some(identifier_packet_fees) = ev.incentivized_packet {
+                let sequence = identifier_packet_fees.packet_id.clone().unwrap().sequence;
+                incentivized_ack_cache
+                    .acquire_write()
+                    .insert(sequence.into(), identifier_packet_fees);
+            }
+        }
+        // If the query failed it could mean that the packet is not incentivized.
+        // The error is logged as debug.
+        Err(e) => {
+            debug!("Query for incentivized packet failed: {e}");
+        }
+    }
+}
+
+/// Using the configured FeesFilter and observed/queried information for
+/// incentivized packets, determine if the SendPacket and WriteAcknowledgement events
+/// should be relayed or not.
+fn filter_batch(
+    batch: EventBatch,
+    incentivized_recv_cache: RwArc<Cache<Sequence, IncentivizedPacket>>,
+    fee_filter: FeesFilters,
+) -> EventBatch {
+    let mut filtered_batch = batch.clone();
```

Comment:
Why not just mutate the `batch` (eg. using `batch.events.retain(|e| ...)`) since it's owned here?